---
title: " Extensions"
output: 
  prettydoc::html_pretty:
    toc: true
    theme: hpstr
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{ggmsa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

Biocpkg <- function(pkg) {
    sprintf("[%s](http://bioconductor.org/packages/%s)", pkg, pkg)
}

CRANpkg <- function(pkg) {
    cran <- "https://cran.r-project.org/package"
    fmt <- "[%s](%s=%s)"
    sprintf(fmt, pkg, cran, pkg)
}

library(ggmsa)
library(ggplot2)

```

#  Functional annotations of ggmsa

ggmsa supports several common functional annotations for MSA. Similar to the ggplot2, we modularize the annotation functionality and users can perform them with `+`.

loading sample data as following:
```{r}
 protein_sequences <- system.file("extdata", "sample.fasta", package = "ggmsa")
 miRNA_sequences <- system.file("extdata", "seedSample.fa", package = "ggmsa")
 nt_sequences <- system.file("extdata", "LeaderRepeat_All.fa", package = "ggmsa")
 tp53_sequences <- system.file("extdata", "tp53.fa", package = "ggmsa")
 tp53_genes <- system.file("extdata", "TP53_genes.xlsx", package = "ggmsa")
```

##  Plot sequence logo `geom_seqlogo()` 

Visualizing multiple sequence alignment with sequence logo.

```{r fig.height=6, fig.width=13, message=FALSE, warning=FALSE}
ggmsa(nt_sequences, font = NULL, color = "Chemistry_NT" ) + geom_seqlogo()
```

```{r fig.height = 3, fig.width = 10, warning=FALSE}
ggmsa(protein_sequences, start = 320, end = 360, char_width = 0.5, color = "Chemistry_AA") +  geom_seqlogo(color = "Chemistry_AA")
```

##  Plot GC content `geom_GC()`

 Visualizing multiple sequence alignment with GC content.

```{r fig.height=6, fig.width=13, message=FALSE, warning=FALSE}
ggmsa(nt_sequences, font = NULL, color = "Chemistry_NT" ) + geom_GC()
```

##  Highlighting the seed in miRNA sequences `geom_seed()`

We provide two methods for highlighting seed sequences.

+ The seed region is shaded

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(miRNA_sequences, font = 'DroidSansMono', none_bg = TRUE) + 
   geom_seed(seed = "GAGGUAG") + theme_void()
```

+ The seed region is highlighted using star

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(miRNA_sequences, font = 'DroidSansMono', color = "Chemistry_NT") + 
   geom_seed(seed = "GAGGUAG", star = TRUE) + theme_void()
```

## Plot sequence conservation bar `geom_msaBar()`
```{r fig.height = 2, fig.width = 10,  message=FALSE, warning = FALSE}
 ggmsa(protein_sequences, 221, 280, font = NULL, seq_name = TRUE) + geom_msaBar()
```

##  Opeaning the Consensus View

Using `consensus_views` option to executes Consensus View. We provide four parameters——`use_dot`,`disagreement`, `ignore_gaps` and `ref` to adjust how the Conesen View executes.

+ `use_dot`: a logical value. Displays characters as dots instead of fading their color in the consensus view.
+ `disagreement`: a logical value. Displays characters that disagreememt to consensus(excludes ambiguous disagreements).
+ `ignore_gaps`: a logical value. When selected TRUE, gaps in column are treated as if that row didn't exist.
+ `ref`: a character string. Specifying the reference sequence which should be one of input sequences.

```{r fig.height = 3, fig.width = 10, warning=FALSE}
 ggmsa(protein_sequences, 300, 350, char_width = 0.5, seq_name = T, consensus_views = T, use_dot = T)
```

```{r fig.height = 3, fig.width = 10, warning=FALSE}
 ggmsa(protein_sequences, 300, 350, char_width = 0.5, seq_name = T, consensus_views = T ,use_dot = T, ref = "PH4H_Rhizobium_loti")
```

##  Breaking Down MSA `facet_msa()`

```{r fig.height = 4, fig.width = 10, message=FALSE,  warning=FALSE}
 # 3 fields
 ggmsa(protein_sequences, end = 300, font = NULL, color = "Chemistry_AA") + facet_msa(field = 100)
```


#  Joint visualization

##  Cross-link with `r Biocpkg("ggtree")` package:

```{r fig.height=2, fig.width=10, message=FALSE, warning=FALSE}
#MSA + tree
library(Biostrings)
x <- readAAStringSet(protein_sequences)
d <- as.dist(stringDist(x, method = "hamming")/width(x)[1])
library(ape)
tree <- bionj(d)
library(ggtree)
p <- ggtree(tree ) + geom_tiplab()

data <- tidy_msa(x, 164, 213)
p + geom_facet(geom = geom_msa, data = data,  panel = 'msa',
               font = NULL, color = "Chemistry_AA") +
    xlim_tree(1)
```

```{r fig.height=2, fig.width=10, message=FALSE, warning=FALSE}
#MSA + sequence logo + tree
library(Biostrings)
x <- readAAStringSet(protein_sequences)
d <- as.dist(stringDist(x, method = "hamming")/width(x)[1])
library(ape)
tree <- bionj(d)
library(ggtree)
p <- ggtree(tree ) + geom_tiplab()

data <- tidy_msa(x, 164, 213)
 p + geom_facet(geom = geom_msa, data = data,
                panel = 'msa', font = NULL,
                color = "Chemistry_AA") +
     geom_facet(geom = ggmsa:::geom_logo, data = data,
               panel = "msa", font = "DroidSansMono",
               color = "Chemistry_AA",adaptive = T) +
               xlim_tree(1)
```


##  Cross-link with `r CRANpkg("gggenes")` package
```{r fig.height=4, fig.width=14, message=FALSE, warning=FALSE}
library(Biostrings)
library(ape)
library(ggtree)
library(gggenes)
library(ggnewscale)
library(RColorBrewer)
#tree
tp53 <- readAAStringSet(tp53_sequences)
d <- as.dist(stringDist(tp53, method = "hamming")/width(tp53)[1])
tree <- bionj(d)
p_tp53 <- ggtree(tree, branch.length = 'none') + geom_tiplab()

#msa
data_53 <- tidy_msa(tp53, start = 0, end = 200)
#genes
TP53_arrow <- readxl::read_xlsx(tp53_genes)
TP53_arrow$direction <- 1
TP53_arrow[TP53_arrow$strand == "reverse","direction"] <- -1

#color
mapping = aes(xmin = start, xmax = end, fill = gene, forward = direction)
my_pal <- colorRampPalette(rev(brewer.pal(n = 10, name = "Set3")))

#tree + genes + msa
p_tp53 + geom_facet(geom = geom_motif,
                    mapping = mapping, data = TP53_arrow,
                    panel = 'genes',  on = 'TP53',
                    arrowhead_height = unit(3, "mm"),
                    arrowhead_width = unit(1, "mm")) +
  scale_fill_manual(values = my_pal(10)) +
  new_scale_fill() +
  geom_facet(geom = geom_msa, data = data_53,
             panel = 'msa', font = NULL,
             color = "Chemistry_AA") + xlim_tree(3.5)
```
