---
title: " Extensions"
output: 
  prettydoc::html_pretty:
    toc: true
    theme: hpstr
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{ggmsa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

Biocpkg <- function(pkg) {
    sprintf("[%s](http://bioconductor.org/packages/%s)", pkg, pkg)
}

library(ggmsa)
library(ggplot2)

```

#  Related Extensions of ggmsa

ggmsa is based on ggplot2, so combining multiple sequence alignment with other plots generated by ggplot2 is available.

loading sample data as following:
```{r}
 protein_sequences <- system.file("extdata", "sample.fasta", package = "ggmsa")
 miRNA_sequences <- system.file("extdata", "seedSample.fa", package = "ggmsa")
 nt_sequences <- system.file("extdata", "LeaderRepeat_All.fa", package = "ggmsa")
 
```


##  Combining with tree `ggtree::geom_facet()`

Cross-link with `r Biocpkg("ggtree")`package:

```{r fig.height=2, fig.width=10, message=FALSE, warning=FALSE}
library(Biostrings)
x <- readAAStringSet(protein_sequences)
d <- as.dist(stringDist(x, method = "hamming")/width(x)[1])
library(ape)
tree <- bionj(d)
library(ggtree)
p <- ggtree(tree ) + geom_tiplab()

data <- tidy_msa(x, 164, 213)
p + geom_facet(geom = geom_msa, data = data,  panel = 'msa',
               font = NULL, color = "Chemistry_AA") +
    xlim_tree(1)
```

##  Plot sequence motifs `geom_seqlogo()` 

Visualizing multiple sequence alignment with sequence logo.

```{r fig.height=6, fig.width=13, message=FALSE, warning=FALSE}
ggmsa(nt_sequences, font = NULL, color = "Chemistry_NT" ) + geom_seqlogo()
```

##  Plot GC content `geom_GC()`

 Visualizing multiple sequence alignment with GC content.

```{r fig.height=6, fig.width=13, message=FALSE, warning=FALSE}
ggmsa(nt_sequences, font = NULL, color = "Chemistry_NT" ) + geom_GC()
```

##  Highlighting the seed in miRNA sequences `geom_seed()`

We provide two methods for highlighting seed sequences.

+ The seed region is shaded

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(miRNA_sequences, font = 'DroidSansMono', none_bg = TRUE) + 
   geom_seed(seed = "GAGGUAG") + theme_void()
```

+ The seed region is highlighted using star

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(miRNA_sequences, font = 'DroidSansMono', color = "Chemistry_NT") + 
   geom_seed(seed = "GAGGUAG", star = TRUE) + theme_void()
```

##  Breaking Down MSA `facet_msa()`

```{r fig.height = 4, fig.width = 10, warning=FALSE}
 # 3 fields
 ggmsa(protein_sequences, end = 300, font = NULL, color = "Chemistry_AA") + facet_msa(field = 100)
```

```{r fig.height = 5, fig.width = 10, warning=FALSE}
 # 4 fields 
 ggmsa(protein_sequences, end = 400, color = "Chemistry_AA") + facet_msa(field = 100)
```
