---
title: "ggmsa: Plot multiple sequence alignment using ggplot2"
output: 
  prettydoc::html_pretty:
    toc: true
    theme: hpstr
date: "`r Sys.Date()`"
bibliography: ggmsa.bib
vignette: >
  %\VignetteIndexEntry{ggmsa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

CRANpkg <- function(pkg) {
    cran <- "https://cran.r-project.org/package"
    fmt <- "[%s](%s=%s)"
    sprintf(fmt, pkg, cran, pkg)
}

Biocpkg <- function(pkg) {
    sprintf("[%s](http://bioconductor.org/packages/%s)", pkg, pkg)
}
```

```{r echo=FALSE, results='hide', message=FALSE}
library(ggmsa)
```

# Abstract

Plot multiple sequence alignment using `r CRANpkg("ggplot2")` with multiple color schemes supported.

# Citation

To do

# Introduction

Supports visualizing multiple sequence alignment of DNA and protein sequences using ggplot2 It supports a number of colour schemes, including Chemistry, Clustal, Shapely, Taylor and Zappo. Multiple sequence alignment can easily be combined with other 'ggplot2' plots, such as aligning a phylogenetic tree produced by 'ggtree' with multiple sequence alignment.

# Installation

This R package (`r CRANpkg("ggmsa")`, current version: `r packageVersion('ggmsa')`) is avalable via CRAN. The way to install the package is the following:

```{r eval=FALSE}
## installing the package
install.packages("ggmsa")
## ## loading the package
library("ggmsa")
```

## Load sample data

The file sample.fasta is shipped with the ggmsa package. To determine where the file is located, enter the following command in your R session:

```{r warning=FALSE}
sequences <- system.file("extdata", "sample.fasta", package = "ggmsa")
print(sequences)
```

# Printing Multiple Sequence Alignments

We offer 5 color schemes for multiple sequence alignments.

## Clustal X Colour Scheme

This is an emulation of the default colourscheme used for alignments in [Clustal X](http://www.jalview.org/help/html/colourSchemes/clustal.html), a graphical interface for the ClustalW multiple sequence alignment program. Each residue in the alignment is assigned a colour if the amino acid profile of the alignment at that position meets some minimum criteria specific for the residue type.

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(sequences, 320, 360, color = "Clustal")
```

## Color by Chemistry

Amino acids are colored according to their side chain chemistry:

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(sequences, 320, 360, color = "Chemistry_AA")
```

## Color by Shapely

This color scheme matches the RasMol amino acid and RasMol nucleotide [color schemes](http://openrasmol.org/doc/rasmol.html#shapelycolours), which are, in turn, based on Robert Fletterickâ€™s "Shapely models".

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(sequences, 320, 360, color = "Shapely_AA")
```

## Color by Taylor

This color scheme is taken from Taylor[@Taylor1997Residual] and is also used in JalView[@Waterhouse2009Jalview].

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(sequences, 320, 360, color = "Taylor_AA")
```

## Color by Zappo

This scheme colors residues according to their physico-chemical properties, and is also used in JalView[@Waterhouse2009Jalview].

```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(sequences, 320, 360, color = "Zappo_AA")
```

## Do not print letters

If you specify font = NULL, only the background box will be printed.
```{r fig.height = 2, fig.width = 10, warning=FALSE}
ggmsa(sequences, 320, 360, font = NULL, color = "Chemistry_AA")
```

# Combining plots

ggmsa is based on ggplot2, so combining multiple sequence alignment with other plots generated by ggplot2 is simple

## Combining with tree

Cross-link with `r Biocpkg("ggtree")`[@yu2017ggtree] package:

```{r fig.height=2, fig.width=10, message=FALSE, warning=FALSE}
library(Biostrings)
x <- readAAStringSet(sequences)
d <- as.dist(stringDist(x, method = "hamming")/width(x)[1])
library(ape)
tree <- bionj(d)
library(ggtree)
p <- ggtree(tree ) + geom_tiplab()

data = tidy_msa(x, 164, 213)
p + geom_facet(geom = geom_msa, data = data,  panel = 'msa',
               font = NULL, color = "Chemistry_AA") +
    xlim_tree(1)
```

## Combining with sequence logo

Cross-link with `r CRANpkg("ggseqlogo")`[@Wagih2017ggseqlogo] package. Bits of sequence letters can be indicated in the sequence logo.

```{r fig.height=6, fig.width=13, message=FALSE, warning=FALSE}
f <- system.file("extdata", "LeaderRepeat_All.fa", package = "ggmsa")
s <- readDNAStringSet(f)
strings <- as.character(s)
p1 <- ggmsa(s, font = NULL, color = 'Chemistry_NT')

library(ggseqlogo)
library(cowplot)

p2 <- axis_canvas(p1, axis='x')+ geom_logo(strings, 'probability') 
pp <- insert_xaxis_grob(p1, p2, position="top", grid::unit(.05, "null"))
ggdraw(pp)
```

# References
